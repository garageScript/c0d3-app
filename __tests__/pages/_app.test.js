import React from 'react'
import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved
} from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import MyApp from '../../pages/_app'
import Login from '../../pages/login'
import * as Sentry from '@sentry/browser'
import GET_APP from '../../graphql/queries/getApp'
import dummyLessonData from '../../__dummy__/lessonData'
import dummySessionData from '../../__dummy__/sessionData'
import dummyAlertData from '../../__dummy__/alertData'
import posthog from 'posthog-js'

jest.mock('posthog-js')
jest.spyOn(Sentry, 'captureException')

const mocks = [
  {
    request: { query: GET_APP },
    result: {
      data: {
        session: dummySessionData,
        lessons: dummyLessonData,
        alerts: dummyAlertData
      }
    }
  }
]

describe('MyApp component', () => {
  const OLD_ENV = process.env
  beforeEach(() => {
    jest.clearAllMocks()
    process.env = { ...OLD_ENV }
  })
  afterEach(() => {
    process.env = OLD_ENV
  })

  test('posthog init function should not be called if not in production environment', async () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MyApp Component={Login} />
      </MockedProvider>
    )
    expect(posthog.init).not.toHaveBeenCalled()
    await waitFor(() =>
      expect(screen.getByRole('heading', { name: /login/i })).toBeTruthy()
    )
  })

  test('posthog init function is being called with the correct arguments', async () => {
    process.env = {
      ...process.env,
      NODE_ENV: 'production',
      POSTHOG_API_KEY: 'fake-posthog-api-key'
    }
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MyApp Component={Login} />
      </MockedProvider>
    )
    expect(posthog.init).toHaveBeenCalledWith(process.env.POSTHOG_API_KEY, {
      api_host: 'https://app.posthog.com'
    })
    await waitFor(() =>
      expect(screen.getByRole('heading', { name: /login/i })).toBeTruthy()
    )
  })

  test('should render Login component passed in as prop', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MyApp Component={Login} />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
    expect(Sentry.captureException).not.toHaveBeenCalled()
    await waitFor(() =>
      expect(screen.getByRole('heading', { name: /login/i })).toBeTruthy()
    )
  })
})
