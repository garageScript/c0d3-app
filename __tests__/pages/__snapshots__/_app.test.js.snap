// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyApp component should render Login component passed in as prop 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        <div
          class="navbar-brand text-primary fw-bold"
        >
          C0D3
        </div>
      </a>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="text-center navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="m-auto navbar-nav"
        >
          <a
            class="nav-item nav-link noUnderline"
            href="/curriculum"
          >
            Curriculum
          </a>
          <a
            class="nav-item nav-link noUnderline"
            href="https://github.com/garageScript/c0d3-app"
            rel="noopener noreferrer"
            target="_blank"
          >
            Repo
          </a>
          <a
            class="nav-item nav-link noUnderline"
            href="/docs/learn"
          >
            Learn
          </a>
          <a
            class="nav-item nav-link noUnderline"
            href="https://discord.gg/c0d3"
            rel="noopener noreferrer"
            target="_blank"
          >
            Community
          </a>
        </div>
        <div
          class="nav-buttons"
        >
          <a
            class="btn btn-light border m-2 me-lg-3 noUnderline"
            href="/login"
          >
            Login
          </a>
          <a
            class="btn btn-light border m-2 me-lg-3 noUnderline"
            href="/signup"
          >
            Signup
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 fw-bold mb-5"
          >
            Login
          </h1>
          <p
            class="card-text"
          />
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="form-group"
            >
              <div
                class="mt-3 d-grid"
              >
                <input
                  class="form-control form-control-lg fw-light mb-3 "
                  data-testid="username"
                  name="username"
                  placeholder="Username"
                  value=""
                />
                <input
                  class="form-control form-control-lg fw-light mb-3 "
                  data-testid="password"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <button
                  class="btn btn-primary btn-lg btn mb-3"
                  data-testid="submit"
                  type="submit"
                >
                  Login to Your Account
                </button>
              </div>
            </div>
          </form>
          <a
            class=" link"
            href="/forgotpassword"
          >
            Forgot your password?
          </a>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 fw-light text-center pb-5 text-muted"
  >
    Â© Copyright 2022 GarageScript. All rights reserved.
  </footer>
</div>
`;
